platform-sdk/swirlds-benchmarks/src/jmh/java/com/swirlds/benchmark/ReconnectBench.java

```diff
@@ -71,62 +71,76 @@ public class ReconnectBench extends VirtualMapBaseBench {
      *
-     * @param map a VirtualMap instance
+     * @param mapRef a reference to a VirtualMap instance
      * @return a populator for the map
      * @param <K> key type
      * @param <V> value type
      */
     private static <K extends VirtualKey, V extends VirtualValue> BiConsumer<K, V> buildVMPopulator(
-            final VirtualMap<K, V> map) {
+            final AtomicReference<VirtualMap<K, V>> mapRef) {
         return (k, v) -> {
             if (v == null) {
-                map.remove(k);
+                mapRef.get().remove(k);
             } else {
-                map.put(k, v);
+                mapRef.get().put(k, v);
             }
         };
     }
 
-    @Setup(Level.Invocation)
-    public void setupInvocation() {
+    /** Generate a state and save it to disk once for the entire benchmark. */
+    @Setup
+    public void setupBenchmark() {
         beforeTest("reconnect");
+        updateMerkleDbPath();
 
-        teacherMap = createEmptyMap("teacher");
-        learnerMap = createEmptyMap("learner");
+        final AtomicReference<VirtualMap<BenchmarkKey, BenchmarkValue>> teacherRef =
+                new AtomicReference<>(createEmptyMap("teacher"));
+        final AtomicReference<VirtualMap<BenchmarkKey, BenchmarkValue>> learnerRef =
+                new AtomicReference<>(createEmptyMap("learner"));
 
         final Random random = new Random(randomSeed);
         new StateBuilder<>(BenchmarkKey::new, BenchmarkValue::new)
                 .buildState(
                         random,
-                        numRecords,
+                        (long) numRecords * (long) numFiles,
                         teacherAddProbability,
                         teacherRemoveProbability,
                         teacherModifyProbability,
-                        buildVMPopulator(teacherMap),
-                        buildVMPopulator(learnerMap));
+                        buildVMPopulator(teacherRef),
+                        buildVMPopulator(learnerRef),
+                        i -> {
+                            if (i % numRecords == 0) {
+                                System.err.printf("Copying files for i = %,d\n", i);
+                                teacherRef.set(copyMap(teacherRef.get()));
+                                learnerRef.set(copyMap(learnerRef.get()));
+                            }
+                        });
+
+        teacherRef.set(flushMap(teacherRef.get()));
+        learnerRef.set(flushMap(learnerRef.get()));
+
+        final List<VirtualMap<BenchmarkKey, BenchmarkValue>> mapCopies =
+                saveMaps(List.of(teacherRef.get(), learnerRef.get()));
+        mapCopies.forEach(this::releaseAndCloseMap);
+    }
 
+    /** Restore the saved state from disk as a new test on-disk copy for each iteration. */
+    @Setup(Level.Invocation)
+    public void setupInvocation() {
+        updateMerkleDbPath();
+
+        teacherMap = restoreMap("teacher");
```

Comment:
`restoreMap()` may return null, if no previously saved state is found... I may be wrong, but looking into code, I'd expect an NPE in this case

Reply:
That's a good catch. I'll handle this (although the bench would still fail with an exception.)

Reply:
Why would it fail with an exception? `restoreMap()` returns null, if there is nothing to restore, i.e. this is the first run without a saved state. This looks like a valid case to me. In this case, a new map should be created

Reply:
This `restoreMap` call occurs in a `@Setup(Level.Invocation)` which runs before every invocation.

However, the benchmark is currently designed to rely on a prior `@Setup public void setupBenchmark()` call, which is implicitly `@Setup(Trial)` that runs before the entire benchmark. That's where the map is being saved. If this method fails, the the benchmarks is supposed to fail because it's not designed to bypass this first @Setup method.